GH-2145: named graph support in federation engine

aschwarte10 opened the pull request:
GH-2145: fix context retrieval in federation engine

* context IDs retrieved from the endpoints have to be materialized as
they are only available while the connection is open
* add unit test infrastructure for named graph tests and define some
test data


GH-2145: add named graph handling for #getStatements API

Add named graph support to the RepositoryConnection API using the
getStatements method. Also covered with various unit tests.

Note that this change provides already the infrastructure to further add
support for named graph queries.


GH-2145: add named graph handling for simple federated SELECT patterns

* support GRAPH clause around single basic graph pattern
* support FROM clause in queries


* add named graph handling for simple federated SELECT patterns
** support GRAPH clause around single basic graph pattern
** support FROM clause in queries

* fix handling of base URIs for query parsing in FedX

* add named graph support for bound joins

GH-2149 GH-2145 use new compliance test suite for testing of FedX

GitHub issue resolved: #2145


---- 
PR Author Checklist: 

 - [ ] my pull request is [self-contained](https://rdf4j.org/documentation/developer/merge-strategy/#self-contained-changes-pull-requests-and-commits)
 - [ ] I've added tests for the changes I made
 - [ ] every commit message starts with the issue number (GH-xxxx) followed by a meaningful description of the change
 - [ ] every commit has been [signed off](https://stackoverflow.com/questions/1962094/what-is-the-sign-off-feature-in-git-for)

Note: we merge all feature pull requests using [squash and merge](https://help.github.com/en/github/administering-a-repository/about-merge-methods-on-github#squashing-your-merge-commits). See [RDF4J git merge strategy](https://rdf4j.org/documentation/developer/merge-strategy/) for more details.

jeenbroekstra commented:
I've also added the generic W3C SPARQL 1.1 compliance tests for FedX. We may end up taking that out again if it gets in the way, but we're currently only failing 5 tests, which is pretty good going.

aschwarte10 commented:
> I've also added the generic W3C SPARQL 1.1 compliance tests for FedX. We may end up taking that out again if it gets in the way, but we're currently only failing 5 tests, which is pretty good going.

Can you shortly comment on how the test data is distributed amongst the federation members?

jeenbroekstra commented:
> > I've also added the generic W3C SPARQL 1.1 compliance tests for FedX. We may end up taking that out again if it gets in the way, but we're currently only failing 5 tests, which is pretty good going.
> 
> Can you shortly comment on how the test data is distributed amongst the federation members?

It currently just uploads all data into the first federation member, which admittedly is not very exciting. But it shouldn't be too hard to tweak the test rig a little so that we can distribute by named graph, so that at least the tests that involve named graphs in their dataset will end up querying multiple members of the federation.

jeenbroekstra commented on tools/federation/src/main/java/org/eclipse/rdf4j/federated/FederationManager.java:
```diff
@@ -332,7 +342,10 @@ public void updateStrategy() {
 
 		if (updated) {
-			strategy = FederationEvaluationStrategyFactory.getEvaluationStrategy(type, federationContext);
+			// TODO problem here: we need to (optionally) inject a Dataset object into the EvaluationStrategy, which is
+			// query-specific and determines what parts of the store are relevant for the query. However this part of
+			// the code assumes that we can instantiate an EvaluationStrategy once and then reuse it for each query.
```
@aschwarte10 I could use your eyes on this problem. FedX is currently designed to instantiate the EvaluationStrategy once (unless the config changes). But in RDF4J Sails, it is generally assumed that each query gets its own EvaluationStategy - that is why the `Dataset` object is passed in on the constructor. I believe that this is caused some of the named graph queries to fail on FedX. I'd like to hear your opinion on the best way forward: can we just remove the assumption that the strategy is defined once and let FedX instantiate a new strategy for each query? 

I'm a little way of doing something like just setting/overwriting the Dataset  assigned to the strategy on each query, as I'm not sure what the consequences would be in a multithreaded environment.

aschwarte10 replied:
Hmm, I see. I wasn't aware of this assumption when initially implementing FedX.

Generally, I do not see a bigger problem in instantiating the strategy for each query (except for the amount of objects). However, quickly scanning the code and the references to where the strategy is passed around, such a change would be a huge refactoring.

I am not sure if at this time right before the release it is such a good idea to do this refactoring. If at all, I would suggest to do it outside of this PR to have a cleaner and easier change.

However, I think we can cope without it: the query specific QueryInfo object always has a reference to the externally passed Dataset. And this QueryInfo is available basically everywhere at evaluation time. This also how all the other named graph queries are working.

Let's synch offline on this and come to the best conclusion.

Note: I will probably re-order your last commit on top (i.e. a force-push), to have some other must changes integrated. The we have the possibility to decide on this special thing

jeenbroekstra replied:
I agree a full refactoring is out of scope (and fwiw, the assumption I'm talking about isn't particularly well documented or obvious, so not blaming your design here). I may have an idea for a quick workaround fix. If I can get that to work I'll push it, otherwise I'll revert the WIP changes I made and log this as a known bug.

jeenbroekstra commented:
Github is reporting test failures but it greenlines when I run it locally.

aschwarte10 commented:
> Github is reporting test failures but it greenlines when I run it locally.
> 
> **edit** I just realized that the tests in FedX are apparently set up to run several times, using different repository configurations. The test succeeds when run on a memory store, but fails on a native store.

Your understanding is right (but the issue is probably unrelated to NativeStore): it is more about the chosen strategy.
Tests are executed three times in different environments

1) Federation members as SPARQL endpoints => SPARQL strategy
2) Federatuin members as RDF4J  HTTP repository => SPARQL strategy
3) Federation members as native store => Sail strategy

The SailStrategy has currently an issue (already identified), the pointer is in https://github.com/eclipse/rdf4j/pull/2146/commits/ed19bbb4e4662535fe682bb2335961ccc0e49019

Once this is addressed the test should also be green.

Note: you can also locally in your IDE switch the environment, see `org.eclipse.rdf4j.federated.SPARQLServerBaseTest.REPOSITORY_TYPE`(i.e. temporarily set it to another value using the enum)

aschwarte10 commented:
We are now in a good shape to get this PR merged. I will clean the commit history and force-push a locally squashed commit containing the entire change.

aschwarte10 commented:
I have cleaned the branch history (squashed and rebased on latest develop). Note that I have split the content into two commits (=> changes to FedX, and changes to compliance test). A backup of the change before the squash is available in https://github.com/aschwarte10/rdf4j/commits/issues/%232145-named-graph-support-backup

The federation tests are now green locally.

Note: the CI check is currently failing due to an independent issue in the MemoryStore module: the issue is available on the develop branch already (maybe introduced to some backmerge from master). Once this is fixed and available in this branch through a rebase, this PR should also be greenlined. For reference: the test is `org.eclipse.rdf4j.sail.memory.QueryPlanRetrievalTest`

@jeenbroekstra maybe you can double check this PR and review.

Open issues for follow ups (which we explicitly do outside this PR):

* fix FROM NAMED handling for SailStrategy (=> #getStatements implementation)
* add support for GRAPH ?g { } (i..e a variable graph)
* Optionally: consider support to deactivate GRAPH handling (i.e. only triple mode)

jeenbroekstra commented:
> I have cleaned the branch history (squashed and rebased on latest develop). 

While I don't mind that you do this, it's not really necessary, as we squash-merge pull requests anyway. 

> Note: the CI check is currently failing due to an independent issue in the MemoryStore module: the issue is available on the develop branch already (maybe introduced to some backmerge from master). Once this is fixed and available in this branch through a rebase, this PR should also be greenlined. For reference: the test is org.eclipse.rdf4j.sail.memory.QueryPlanRetrievalTest

Ah thanks. I have a fix for that lined up, will merge into this branch as soon as it's available on develop.

jeenbroekstra commented:
I've logged issues for the two non-optional things you mentioned: #2185 and #2186. I'd like to sleep a little on whether making graph support optional is something that adds value.
